src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -6,193 +6,592 @@
+
+        /// <summary>
+        /// Write a <see cref="Guid"/> value into the current array.
+        /// </summary>
+        /// <param name="value">The <see cref="Guid"/> value to be written to JSON data.</param>
+        public void WriteValue(Guid value)
         {
-            ConsideComma();
-            WriteIndent();
-            _formatter.Append('"');
-            _formatter.Append(name);
-            _formatter.Append("\":");
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteGuid(value);
         }
 
-        private void ConsideComma()
+        /// <summary>
+        /// Write a null value into the current array.
+        /// </summary>
+        public void WriteNull()
         {
-            if (_wroteListItem) {
-                _formatter.Append(',');
-                WriteNewline();
-            }
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteAscii(JsonConstants.NullValue);
+        }
+
+        /// <summary>
+        /// Write an undefined value into the current array.
+        /// </summary>
+        public void WriteUndefined()
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+            WriteSpacing();
+            WriteAscii(JsonConstants.UndefinedValue);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteStartAttribute(string name)
+        {
+            WriteItemSeperator();
+            _firstItem = false;
+
+            WriteSpacing();
+            WriteQuotedString(name);
+            WriteChar(JsonConstants.KeyValueSeperator);
+
+            if (_prettyPrint)
+                WriteChar(JsonConstants.Space);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void WriteNewline()
+        private void WriteChar(char value)
         {
-            if (PrettyPrint) {
-                _formatter.Append('\n');
+            var output = _output;
+            if (UseFastUtf8)
+            {
+                if (output.Buffer.Length < 1)
+                    output.Enlarge();
+
+                output.Buffer.DangerousGetPinnableReference() = (byte)value;
+                output.Advance(1);
             }
+            else if (UseFastUtf16)
+            {
+                if (output.Buffer.Length < 2)
+                    output.Enlarge(2);
+
+                unsafe
+                {
+                    Unsafe.AsRef<char>(Unsafe.AsPointer<byte>(ref output.Buffer.DangerousGetPinnableReference())) = value;
+                }
+
+                output.Advance(2);
+            }
+            else
+            {
+                unsafe
+                {
+                    // Slow path, if we are dealing with non-invariant.
+                    ReadOnlySpan<char> chSpan = new ReadOnlySpan<char>(&value, 1);
+                    Write(chSpan);
```

Comment:
For reference, I disabled the shortcuts and used the encoder I was given, effectively taking all the slow paths for both UTF-8 and UTF-16. Below you can see the comparison. In the formatted scenarios, it makes a big difference. In the non-formatted scenarios, it makes much less difference.

| Test | New | JSON.NET | JSON.NET / New |
|---|---|---|---|
| Json Writer (True, Utf8) | 1825 | 1951 | 0.935 |
| Json Writer (False, Utf8) | 1499 | 1357 | 1.105 |
| Json Writer (True, Utf16) | 1845 | 1914 | 0.964 |
| Json Writer (False, Utf16) | 1511 | 1349 | 1.120 |
| Json Writer (True, Utf8-Slow) | 4197 | 2066 | 2.031 |
| Json Writer (False, Utf8-Slow) | 1608 | 1456 | 1.104 |
| Json Writer (True, Utf16-Slow) | 4415 | 2057 | 2.146 |
| Json Writer (False, Utf16-Slow) | 1646 | 1431 | 1.150 |